pkg/util/atomic/atomicrunstate.go

```diff
@@ -0,0 +1,87 @@
+package atomic
+
+import (
+	"sync"
+)
+
+// AtomicRunState can be used to provide thread-safe start/stop functionality to internal run-loops
+// inside a goroutine.
+type AtomicRunState struct {
+	lock     sync.Mutex
+	stopping bool
+	stop     chan struct{}
+	reset    chan struct{}
+}
+
+// Start checks for an existing run state and returns false if the run state has already started. If
+// the run state has not started, then it will advance to the started state and return true.
+func (ars *AtomicRunState) Start() bool {
+	ars.lock.Lock()
+	defer ars.lock.Unlock()
+
+	if ars.stop != nil {
+		return false
+	}
+
+	ars.stop = make(chan struct{}, 1)
```

Comment:
Since, as far as I can tell, signaling only happens by closing `ars.stop` and nothing is ever sent on the channel, this could just be:
```suggestion
	ars.stop = make(chan struct{}, 0)
```